tf/edgeml/trainer/bonsaiTrainer.py

```diff
@@ -0,0 +1,352 @@
+import tensorflow as tf
+import edgeml.utils as utils
+import numpy as np
+import os
+import sys
+
+
+class BonsaiTrainer:
+    def __init__(self, bonsaiObj, lW, lT, lV, lZ, sW, sT, sV, sZ,
+                 learningRate, X, Y, useMCHLoss=True):
+        '''
+        bonsaiObj - Initialised Bonsai Object and Graph
+        lW, lT, lV and lZ are regularisers to Bonsai Params
+        sW, sT, sV and sZ are sparsity factors to Bonsai Params
+        learningRate - learningRate fro optimizer
+        X is the Data Placeholder - Dims [_, dataDimension]
+        Y - Label placeholder for loss computation
+        useMCHLoss - For choice between HingeLoss vs CrossEntropy
+        useMCHLoss - True - MultiClass - multiClassHingeLoss
+        useMCHLoss - False - MultiClass - crossEntropyLoss
+        '''
+
+        self.bonsaiObj = bonsaiObj
+
+        self.lW = lW
+        self.lV = lV
+        self.lT = lT
+        self.lZ = lZ
+
+        self.sW = sW
+        self.sV = sV
+        self.sT = sT
+        self.sZ = sZ
+
+        self.Y = Y
+        self.X = X
+
+        self.useMCHLoss = useMCHLoss
+
+        self.learningRate = learningRate
+
+        self.assertInit()
+
+        self.score, self.X_ = self.bonsaiObj(self.X)
```

Comment:
@adityakusupati  I'm unsure how `tf`  will behave in the following case. Do you have an idea?

1. `bonsaiObj.__call__` was already called outside before passing to the trainer
2. `bonsaiTrainer` is called again on the same object.

In both cases, the first call to `__call__` will create the graph, but I'm unsure what happens when the call happens again (after the graph is already created).

Reply:
I think it re-writes the graph entirely as there are variables such as self.nodeProb etc., inherent to the object and they will be re-created. The old graph will not make any sense even though it is present in the current session as it has no binding even if it is present (ie., no path for backprop). I don't think we should worry about it as BonsaiTrainer is just an example routine (which is same as the one we used ICML). There are various other training routines like DSD which people can custom implement.

Try not to over think these. We are not tailor making EdgeML for every single purpose.

Reply:
I agree. I just was not sure how multiple graph calls interact with tf. If you think this is fine, i'm okay.